## 数据说明与切分要求


**数据集说明**：这份数据集是金融数据（非原始数据，已经处理过了），我们要做的是预测贷款用户是否会逾期。表格中 "status" 是结果标签：0表示未逾期，1表示逾期。

**切分要求**：数据切分方式 - 三七分，其中测试集30%，训练集70%，随机种子设置为2018

## 技术路线

任务一 数据分析

**数据探索**

* 确定df的shape，概览df的columns，df.info，概览df的统计信息
* 确定df的dtype和相应的字段
* 删除全部值相同的列

**数据分析**
* 进行单变量分析
    * 连续变量：集中趋势度量，离中趋势度量（df.descibe+range极差，var变异系数，四分位数间距），箱型图，直方图
    * 分类变量：集中趋势度量（df.descibe+df.mode，比例？），柱形图
* 区分变量类型
* 双变量分析
    * 连续变量和连续变量：√散点图（根据散点图的形状能够显示(indicates )出变量是线性(linear )关系还是非线性(non-linear )的关系）
    * 分类变量和分类变量：√双向表
    * 类别变量和连续变量：当处理类别型和连续性变量的时候，可以为每一个类别变量绘制箱线图
* 时间序列分析

**缺失值处理**
* 缺失值的处理包括：删除，填补，不处理。
* 一般原则：结合实际业务，如果缺失的数据对预测数据无影响或影响较小，则缺失率 > 80%，删除特征/记录；如果确实的数据对预测数据影响较大，且可以通过其业务意义或其他途径补全，则保留。
* 对于待填补的数据

**异常值检测和处理**



任务二 特征工程

## 目录

1. [任务1 数据分析](#任务1)
2. [任务2 特征工程](#任务2)

<a id='任务1'></a>
## 任务1 数据分析

任务1内容：对数据进行探索和分析，包括：

* 数据类型的分析
* 无关特征删除
* 数据类型转换
* 缺失值处理

……以及你能想到和借鉴的数据分析处理

DDL：20190806 10:00pm



In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import sklearn as sk
import datetime
import os
import seaborn as sns

In [2]:
# 设置显示行数
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)
# 限制列宽，每列最多显示15个字符
# pd.set_option('display.max_colwidth', 15)

In [3]:
# 读取data.csv文件
df = pd.read_csv('./data.csv', encoding='gbk')

### 1.1 数据探索

In [4]:
df.head().T

0  \
Unnamed: 0                                                                 5   
custid                                                               2791858   
trade_no                                    20180507115231274000000023057383   
bank_card_no                                                             卡号1   
low_volume_percent                                                      0.01   
middle_volume_percent                                                   0.99   
take_amount_in_later_12_month_highest                                      0   
trans_amount_increase_rate_lately                                        0.9   
trans_activity_month                                                    0.55   
trans_activity_day                                                     0.313   
transd_mcc                                                                17   
trans_days_interval_filter                                                27   
trans_days_interval                                                       26   
regional_mobility                                                          3   
student_feature                                                          NaN   
repayment_capability                                                   19890   
is_high_user                                                               0   
number_of_trans_from_2011                                                 30   
first_transaction_time                                           2.01308e+07   
historical_trans_amount                                               149050   
historical_trans_day                                                     151   
rank_trad_1_month                                                        0.4   
trans_amount_3_month                                                   34030   
avg_consume_less_12_valid_month                                            7   
abs                                                                     3920   
top_trans_count_last_1_month                                            0.15   
avg_price_last_12_month                                                 1020   
avg_price_top_last_12_valid_month                                       0.55   
reg_preference_for_trad                                                 一线城市   
trans_top_time_last_1_month                                                4   
trans_top_time_last_6_month                                               19   
consume_top_time_last_1_month                                              4   
consume_top_time_last_6_month                                             19   
cross_consume_count_last_1_month                                           1   
trans_fail_top_count_enum_last_1_month                                     1   
trans_fail_top_count_enum_last_6_month                                     2   
trans_fail_top_count_enum_last_12_month                                    2   
consume_mini_time_last_1_month                                             5   
max_cumulative_consume_later_1_month                                    2170   
max_consume_count_later_6_month                                            6   
railway_consume_count_last_12_month                                        0   
pawns_auctions_trusts_consume_last_1_month                              1970   
pawns_auctions_trusts_consume_last_6_month                             18040   
jewelry_consume_count_last_6_month                                         0   
status                                                                     1   
source                                                                    xs   
first_transaction_day                                                   1738   
trans_day_last_12_month                                                   85   
id_name                                                                   蒋红   
apply_score                                                              58

In [5]:
# 查看数据集形状
df.shape

(4754, 90)

In [6]:
# 查看数据信息
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4754 entries, 0 to 4753
Data columns (total 90 columns):
Unnamed: 0                                    4754 non-null int64
custid                                        4754 non-null int64
trade_no                                      4754 non-null object
bank_card_no                                  4754 non-null object
low_volume_percent                            4752 non-null float64
middle_volume_percent                         4752 non-null float64
take_amount_in_later_12_month_highest         4754 non-null int64
trans_amount_increase_rate_lately             4751 non-null float64
trans_activity_month                          4752 non-null float64
trans_activity_day                            4752 non-null float64
transd_mcc                                    4752 non-null float64
trans_days_interval_filter                    4746 non-null float64
trans_days_interval                           4752 non-null float64
regional_mobility

In [7]:
# 查看dtype
def dtype_clasf(df):
    typedic= {} # 类型字典
    for name in df.columns:
        typedic[str(df[name].dtype)] = typedic.get(str(df[name].dtype),[])+[name]
    for key,value in typedic.items():
        # print('we have {} columns in type {}'.format(len(value),key))
        print('There are {} columns in dtype {}, they are {} \n'.format(len(value),key,value))

In [8]:
dtype_clasf(df)

There are 13 columns in dtype int64, they are ['Unnamed: 0', 'custid', 'take_amount_in_later_12_month_highest', 'repayment_capability', 'is_high_user', 'historical_trans_amount', 'trans_amount_3_month', 'abs', 'avg_price_last_12_month', 'max_cumulative_consume_later_1_month', 'pawns_auctions_trusts_consume_last_1_month', 'pawns_auctions_trusts_consume_last_6_month', 'status'] 

There are 7 columns in dtype object, they are ['trade_no', 'bank_card_no', 'reg_preference_for_trad', 'source', 'id_name', 'latest_query_time', 'loans_latest_time'] 

There are 70 columns in dtype float64, they are ['low_volume_percent', 'middle_volume_percent', 'trans_amount_increase_rate_lately', 'trans_activity_month', 'trans_activity_day', 'transd_mcc', 'trans_days_interval_filter', 'trans_days_interval', 'regional_mobility', 'student_feature', 'number_of_trans_from_2011', 'first_transaction_time', 'historical_trans_day', 'rank_trad_1_month', 'avg_consume_less_12_valid_month', 'top_trans_count_last_1_month',

In [9]:
# 查看数据统计信息
df.describe()

Unnamed: 0        custid  low_volume_percent  middle_volume_percent  \
count   4754.000000  4.754000e+03         4752.000000            4752.000000   
mean    6008.414178  1.690993e+06            0.021806               0.901294   
std     3452.071428  1.034235e+06            0.041527               0.144856   
min        5.000000  1.140000e+02            0.000000               0.000000   
25%     3106.000000  7.593358e+05            0.010000               0.880000   
50%     6006.500000  1.634942e+06            0.010000               0.960000   
75%     8999.000000  2.597905e+06            0.020000               0.990000   
max    11992.000000  4.004694e+06            1.000000               1.000000   

       take_amount_in_later_12_month_highest  \
count                            4754.000000   
mean                             1940.197728   
std                              3923.971494   
min                                 0.000000   
25%                                 0.000000   
50%                               500.000000   
75%                              2000.000000   
max                             68000.000000   

       trans_amount_increase_rate_lately  trans_activity_month  \
count                        4751.000000           4752.000000   
mean                           14.160674              0.804411   
std                           694.180473              0.196920   
min                             0.000000              0.120000   
25%                             0.615000              0.670000   
50%                             0.970000              0.860000   
75%                             1.600000              1.000000   
max                         47596.740000              1.000000   

       trans_activity_day   transd_mcc  trans_days_interval_filter  \
count         4752.000000  4752.000000                 4746.000000   
mean             0.365425    17.502946                   29.029920   
std              0.170196     4.475616                   22.722432   
min              0.033000     2.000000                    0.000000   
25%              0.233000    15.000000                   16.000000   
50%              0.350000    17.000000                   23.000000   
75%              0.480000    20.000000                   32.000000   
max              0.941000    42.000000                  285.000000   

       trans_days_interval  regional_mobility  student_feature  \
count          4752.000000        4752.000000      1756.000000   
mean             21.751263           2.678662         1.001139   
std              16.474916           0.890360         0.033739   
min               4.000000           1.000000         1.000000   
25%              12.000000           2.000000         1.000000   
50%              17.000000           3.000000         1.000000   
75%              27.000000           3.000000         1.000000   
max             234.000000           5.000000         2.000000   

       repayment_capability  is_high_user  number_of_trans_from_2011  \
count          4.754000e+03   4754.000000                4752.000000   
mean           1.870201e+04      0.011149                  23.033880   
std            5.221783e+04      0.105007                  10.057837   
min            0.000000e+00      0.000000                   1.000000   
25%            8.590000e+03      0.000000                  16.000000   
50%            1.221000e+04      0.000000                  21.000000   
75%            1.764750e+04      0.000000                  29.000000   
max            2.459390e+06      1.000000                  85.000000   

       first_transaction_time  historical_trans_amount  historical_trans_day  \
count            4.752000e+03             4.754000e+03           4752.000000   
mean             2.015109e+07             2.307359e+05            176.109428   
std              1.480487e+04             3.204931e+05             99.687285   
min              2.011010e+07             0.000000e+00           

In [13]:
# 查看数据缺失率
missing_value_fri = (df.isnull().sum()/df.isnull().count()).sort_values(ascending=True)
missing_value_fri

Unnamed: 0                                    0.000000
source                                        0.000000
pawns_auctions_trusts_consume_last_6_month    0.000000
pawns_auctions_trusts_consume_last_1_month    0.000000
max_cumulative_consume_later_1_month          0.000000
avg_price_last_12_month                       0.000000
abs                                           0.000000
trans_amount_3_month                          0.000000
historical_trans_amount                       0.000000
is_high_user                                  0.000000
repayment_capability                          0.000000
status                                        0.000000
trade_no                                      0.000000
custid                                        0.000000
take_amount_in_later_12_month_highest         0.000000
bank_card_no                                  0.000000
trans_days_interval                           0.000421
first_transaction_day                         0.000421
reg_prefer

In [25]:
missing_value_fri.value_counts()

0.062474    30
0.000421    16
0.000000    16
0.063946    11
0.001683     6
0.003366     3
0.002524     2
0.630627     1
0.089609     1
0.005469     1
0.000631     1
0.058056     1
0.021876     1
dtype: int64

**小结**：

1. 原数据集有4754行数据，90个特征

2. 原数据集的特征中，有70个特征值的数据类型是float64，13个是int64，7个object

3. 其中，数据类型为object的特征包括：
    * 'trade_no'
    * 'bank_card_no'
    * 'reg_preference_for_trad'
    * 'source'
    * 'id_name'
    * 'latest_query_time'
    * 'loans_latest_time'

4. 大部分特征有缺失值，缺失率最高的是student_feature，达到63%；其次是cross_consume_count_last_1_month，为9%；超过二分之一的特征缺失值达到6%以上。

### 1.2 数据分析

### 处理缺失值

缺失值的处理方法为：

* 删除，或
* 填补，或
* 不处理

一般采用删除或填补的方法，首选基于业务的填补方法，其次根据单变量分析进行填补

缺失值处理基本原则：当缺失值大于80%时，考虑删除该行/列


In [ ]:
# 删除缺失值达到80%的行
df1 = df1.dropna(thresh=65,axis=0)

In [ ]:
df1.shape

In [ ]:
# 查看df1各特征缺失值占比
missing_value_fri = (df1.isnull().sum()/df1.isnull().count()).sort_values(ascending=True)
missing_value_fri

由df1缺失值分析可知，student_feature特征缺失值最多，达到62.7%；其次是cross_consume_count_last_1_month，缺失值达到8.7%；avg_price_top_last_12_valid_month的缺失值为2.2%左右。其他的特征缺失值在1%以下或者无缺失值。

In [ ]:
# 分组查看student_feature列的值
df1['student_feature'].value_counts()

In [ ]:
# student_feature列含1和2两个值，缺失值占63%左右
# 在实际业务中，银行会在客户贷款前得知客户的职业。
# 由此推测，student_feature值为空代表该账户非学生账户（不适用），值为2属于错误数据
# 对于student_feature列，将NA填充为0，代表非学生，2以众数0替代
df1['student_feature'] = df1['student_feature'].fillna(0)
df1['student_feature'] = df1['student_feature'].replace([2],[0])

In [ ]:
# 查看数据类型
df1.dtypes.value_counts()

In [ ]:
df1.dtypes

<a id='任务2'></a>
## 任务2 特征工程

任务2内容包括：

* 特征衍生
* 特征挑选：分别用IV值和随机森林等进行特征选择

……以及你能想到特征工程处理

DDL：20190809 8:00pm